lib/spack/spack/build_systems/cuda.py

```diff
@@ -44,23 +42,75 @@ def cuda_flags(arch_list):
+    conflicts('%gcc@7:', when='+cuda ^cuda@:9.1' + arch_platform)
+    conflicts('%gcc@8:', when='+cuda ^cuda@:10' + arch_platform)
+    conflicts('%pgi@:14.8', when='+cuda ^cuda@:7.0.27' + arch_platform)
+    conflicts('%pgi@:15.3,15.5:', when='+cuda ^cuda@7.5' + arch_platform)
+    conflicts('%pgi@:16.2,16.0:16.3', when='+cuda ^cuda@8' + arch_platform)
+    conflicts('%pgi@:15,18:', when='+cuda ^cuda@9.0:9.1' + arch_platform)
+    conflicts('%pgi@:16', when='+cuda ^cuda@9.2.88:10' + arch_platform)
+    conflicts('%clang@:3.5', when='+cuda ^cuda@:7.5' + arch_platform)
+    conflicts('%clang@:3.7,4:',
+              when='+cuda ^cuda@9.0.69:9.0.176' + arch_platform)
+    conflicts('%clang@:3.7,4.1:',
+              when='+cuda ^cuda@9.1.85:9.1' + arch_platform)
+    conflicts('%clang@:3.7,5.1:', when='+cuda ^cuda@9.2' + arch_platform)
+    conflicts('%clang@:3.7,6.1:', when='+cuda ^cuda@:10' + arch_platform)
+
+    # x86_64 vs. ppc64le differ according to NVidia docs
+    # Linux ppc64le compiler conflicts from Table from the docs below:
+    # https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html
+    # https://docs.nvidia.com/cuda/archive/9.2/cuda-installation-guide-linux/index.html
+    # https://docs.nvidia.com/cuda/archive/9.1/cuda-installation-guide-linux/index.html
+    # https://docs.nvidia.com/cuda/archive/9.0/cuda-installation-guide-linux/index.html
+    # https://docs.nvidia.com/cuda/archive/8.0/cuda-installation-guide-linux/index.html
+
+    arch_platform = ' arch=ppc64le platform=linux'
+    # information prior to CUDA 9 difficult to find
+    conflicts('%gcc@6:', when='+cuda ^cuda@:9' + arch_platform)
+    conflicts('%gcc@8:', when='+cuda ^cuda@:10' + arch_platform)
+    conflicts('%pgi', when='+cuda ^cuda@:8' + arch_platform)
+    conflicts('%pgi@:16', when='+cuda ^cuda@:9.1.185' + arch_platform)
+    conflicts('%pgi@:17', when='+cuda ^cuda@:10' + arch_platform)
+    conflicts('%clang@4:', when='+cuda ^cuda@:9.0.176' + arch_platform)
+    conflicts('%clang@5:', when='+cuda ^cuda@:9.1' + arch_platform)
+    conflicts('%clang@6:', when='+cuda ^cuda@:9.2' + arch_platform)
+    conflicts('%clang@7:', when='+cuda ^cuda@:10' + arch_platform)
+
+    # Intel is mostly relevant for x86_64 Linux, even though it also
+    # exists for Mac OS X.
     conflicts('%intel@:14,16:', when='+cuda ^cuda@7.5')
     conflicts('%intel@:14,17:', when='+cuda ^cuda@8.0.44')
-    conflicts('%intel@:14,18:', when='+cuda ^cuda@8.0.61:9')
+    conflicts('%intel@:14,18:', when='+cuda ^cuda@8.0.61:9.1')
+    conflicts('%intel@17:18:', when='+cuda ^cuda@9.2:')
+    conflicts('%intel@19:', when='+cuda')
+
+    # XL is mostly relevant for ppc64le Linux
+    conflicts('%xl@:12,14:', when='+cuda ^cuda@:9.1')
+    conflicts('%xl@:12,14:15', when='+cuda ^cuda@9.2:10')
+    conflicts('%xl@17:', when='+cuda ^cuda ^cuda@:10')
+
+    # Mac OS X
+    platform = ' platform=darwin'
+    conflicts('%clang@:3.4,3.7:', when='+cuda ^cuda@7.5' + platform)
```

Comment:
Since Apple‘s clang in Spack has the form ´clang@7.3.0-apple‘, this and next few lines will prevent using CUDA with Apple‘s clang.

I wish the versioning convention was chosen as ˋ@apple-x.y.zˋ, then it would be easy to distinguish between two families.

Reply:
@davydden I think that I really messed up the MacOS X. Do you have a table like https://gist.github.com/ax3l/9489132. Also, what are the other relevant compilers for MacOS X and CUDA? In principle wouldn't all the `x86_64 Linux` compiler conflicts apply as well?

I would agree that making it easier to distinguish between the Apple-provided clang (from Xcode) versus the clang from github would make life easier. If we are going to change the name convention, why not do: 'clang-apple@x.y.z' ?

Reply:
The only officially supported host-side ccompiler for NVCC is AppleClang in XCode:
https://docs.nvidia.com/cuda/cuda-installation-guide-mac-os-x/index.html#abstract

Unofficially, more might work (such as vanilla clang and gcc)...

Reply:
> If we are going to change the name convention, why not do: 'clang-apple@x.y.z' ?

That’s also an option but IMO is more intrusive. Also we should not call compiler wrappers clang-apple as there are packages that might rely on naming. 

The way versions of Apple’s clang are defined most likely is relevant only for conflicts and alike. I just hope that if we are to adopt ‘@apple-x.y.z’ very few packages will be affected.

Another approach is to extend conflicts so that one could write more complicated conditions, but I think this is an overkill. By simply prefixing the Apple’s versions we would be able to distinguish between the two.

Reply:
If I follow the conversation correctly, is the main problem with current form ´clang@7.3.0-apple‘ that you cannot specify index ranges properly?

Reply:
> , is the main problem with current form ´clang@7.3.0-apple‘ that you cannot specify index ranges properly?

Exsctly, that’s why we had no constraints for clang on Darwin. 

To get this PR going, I would keep it this easy until Spack is more flexible in this. We can also assume for now that on Darein clang is Apple’s.